lib/assert.js

```diff
@@ -931,6 +933,87 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+class Tracker {
+
+  constructor() {
+    this.callChecks = new SafeSet();
+  }
+
+  calls(fn, exact = 1) {
+    if (process._exiting)
+      throw new ERR_CALLS_UNAVAILABLE_DURING_EXIT();
+    if (typeof fn === 'number') {
+      exact = fn;
+      fn = noop;
+    } else if (fn === undefined) {
+      fn = noop;
+    }
+
+    validateUint32(exact, 'exact', true);
+
+    const context = {
+      exact,
+      actual: 0,
+      // eslint-disable-next-line no-restricted-syntax
+      stackTrace: new Error(),
+      name: fn.name || 'calls'
+    };
+    this.callChecks.add(context);
+    const callChecks = this.callChecks;
+
+    return function() {
+      // Increment call count
+      context.actual++;
+      if (context.actual === context.exact) {
+        // Once function has reached its call count remove it from
+        // callChecks set to prevent memory leaks
+        callChecks.delete(context);
+      }
+      if (context.actual > context.exact) {
+        // If call count exceeds exact throw an error
+        const message = `Expected the ${context.name} function to be ` +
+        `executed ${context.exact} time(s) but was executed ` +
+        `${context.actual} time(s).`;
+
+        throw new AssertionError({
+          actual: context.actual,
+          expected: context.exact,
+          message,
+          operator: context.name,
+        });
+      }
+      return fn.apply(this, arguments);
+    };
+  }
+
+  verify() {
+    for (const context of this.callChecks) {
+      // const console = require('console');
+      // console.log(context);
+      if (context.actual !== context.exact) {
+        const message = `Expected the ${context.name} function to be ` +
+        `executed ${context.exact} time(s) but was executed ` +
+        `${context.actual} time(s).`;
+        const console = require('console');
+        // Print stack trace of the function
+        console.error(context.stackTrace);
+        throw new AssertionError({
+          actual: context.actual,
+          expected: context.exact,
+          message,
+          operator: context.name,
+        });
+      }
+    }
+  }
```

Comment:
By throwing the assertion error within the for loop we end up only being notified of the first failure, hiding the others that may be potentially there. I think it would be most helpful to report on all the failures at once... so something like this would be better:

```suggestion
    let detail = [];
    for (const context of this.callChecks) {
      // const console = require('console');
      // console.log(context);
      if (context.actual !== context.exact) {
        const message = `Expected the ${context.name} function to be ` +
        `executed ${context.exact} time(s) but was executed ` +
        `${context.actual} time(s).`;
        detail.push({
          message,
          actual: context.actual,
          expected: context.expected,
          operator: context.name,
          stackTrace: context.stackTrace
        });
      }
    }
    if (detail.length > 0)
      throw new AssertionError({ detail });
  }
```

Thinking about it now, I also think separating out the console.error reporting from the verify would be good by introducing a separate report() function like:

```
report() {
  try {
    this.verify();
  } catch (err) {
    if (err.detail) {
      for (const item of detail) {
        console.error(item.message);
        console.error(item.stackTrace);
      }
    }
  }
}
```

Reply:
Do you suggest I make a new error that can accept an arrays? Or is there some functionality of AssertionErrors that I'm missing

Reply:
As luck would have it TC39 is actually working on adding a built in error type that can aggregate multiple errors but it'll be a while until we get it. For now, I would just throw a single error with the `detail` property like I illustrate above.